### ISSUE TITLE
Relying on modification time may lead to unexpected results (issues with backup-list, and backup-fetch LATEST)

### ISSUE THREAD

**Issue by NikolayS** [wal-g/wal-g#694]
As I can see (from https://github.com/wal-g/wal-g/blob/master/internal/delete_handler.go#L92, https://github.com/wal-g/wal-g/blob/master/internal/delete_handler.go#L47) modification time is considered as the parameter to order full backups.

This may lead to undesired results when we list full backups or, which is even worse and more important, want to restore from the LATEST full backup.

Two examples when using modification time may cause issues

1. Race condition in the case of overlapping `backup-push`:
    - we start `backup-push` on one node of a 10 TiB database. Let's name this backup, say, A;
    - 1 hour later we start `backup-push` on another node, with better disks – backup B;
    - B is finished sooner than A, therefore it has older modification time;
    - as a result, `backup-list` shows B as older, which is not true, and `backup-fetch LATEST` will work with A, not B, which is also a mistake.
1. If some external tool does something with old backups, the modification time is changed. I understand that WAL-G doesn't expect this, but this is exactly what may easily happen (see https://gitlab.com/gitlab-com/gl-infra/production/-/issues/2297#note_368324826 – old backups are moved to colder/cheaper storage, this action changes their modification time, and as a result, `backup-list` output looks messy, and `backup-fetch LATEST` is not reliable anymore, it cannot be used because it may produce unpredicted results).

Note that WAL-E `backup-list` orders backups differently: it uses LSNs to order them (however, it also shows modification time).

I propose using either LSNs or creation time and considering modification time only as an additional data, which by no means may be used when making decisions which one of the available full backups is the "LATEST".

**Comment by NikolayS** [comment]
I realized, that technically, in the example above with backups A and B, A (which started earlier but finished later than B) may be considered as a "fresher" one because when we restore from it, we replay more WALs and achieve a fresher state. However, still, I would choose B because it would allow me to fetch fewer data from the archive and restore faster -- additional WALs I would replay with `wal-fetch`.

If the speed is the same for both nodes or B is slower, ordering by creation time and modification time would lead the same results (first A, then B, so B is the newest).

In other words, creation time looks more attractive in any case.

**Comment by NikolayS** [comment]
cc @x4m @g0djan

**Comment by eReGeBe** [comment]
@x4m checking in on this issue for your input.

**Comment by NikolayS** [comment]
Testing 1.1 that should have this issue fixed. 

What we have:
- daily full backups stored in GCs
- older backups automatically moved to a "colder" storage, which updates their modtime, so they pop up in the tail

1.1 doesn't seem to have this problem fixed:
```
$ sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g --version
wal-g version v1.1	cadf598	2021.08.17_18:17:11	PostgreSQL
$ sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list | tail -10
base_000000050005B94000000022 2021-09-19T10:27:44Z 000000050005B94000000022
base_000000050005BB4C000000B5 2021-09-20T09:37:42Z 000000050005BB4C000000B5
base_000000050005E9DB000000EE 2021-09-20T10:08:28Z 000000050005E9DB000000EE
base_0000000500060413000000F9 2021-09-20T12:41:36Z 0000000500060413000000F9
base_000000050006063800000080 2021-09-21T08:30:01Z 000000050006063800000080
base_000000050005BD6A00000000 2021-09-21T10:12:21Z 000000050005BD6A00000000
base_000000050005EBB70000000F 2021-09-21T20:33:29Z 000000050005EBB70000000F
base_0000000500060895000000C3 2021-09-22T08:38:31Z 0000000500060895000000C3
base_000000050005EE1000000049 2021-09-22T15:30:13Z 000000050005EE1000000049
base_000000050005BF8900000015 2021-09-22T17:39:19Z 000000050005BF8900000015
```


I also see that the output of `backup-list` shows `modified` in the second column, not `created` – this seems to be inconsistent with the supposed order.

We need to ignore modtime and use either LSN or creation time, order by it, and present in the output explicitly.

**Comment by NikolayS** [comment]
cc @x4m

**Comment by usernamedt** [comment]
> Testing 1.1 that should have this issue fixed.
> 
> What we have:
> 
> * daily full backups stored in GCs
> * older backups automatically moved to a "colder" storage, which updates their modtime, so they pop up in the tail
> 
> 1.1 doesn't seem to have this problem fixed:
> 
> ```
> $ sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g --version
> wal-g version v1.1	cadf598	2021.08.17_18:17:11	PostgreSQL
> $ sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list | tail -10
> base_000000050005B94000000022 2021-09-19T10:27:44Z 000000050005B94000000022
> base_000000050005BB4C000000B5 2021-09-20T09:37:42Z 000000050005BB4C000000B5
> base_000000050005E9DB000000EE 2021-09-20T10:08:28Z 000000050005E9DB000000EE
> base_0000000500060413000000F9 2021-09-20T12:41:36Z 0000000500060413000000F9
> base_000000050006063800000080 2021-09-21T08:30:01Z 000000050006063800000080
> base_000000050005BD6A00000000 2021-09-21T10:12:21Z 000000050005BD6A00000000
> base_000000050005EBB70000000F 2021-09-21T20:33:29Z 000000050005EBB70000000F
> base_0000000500060895000000C3 2021-09-22T08:38:31Z 0000000500060895000000C3
> base_000000050005EE1000000049 2021-09-22T15:30:13Z 000000050005EE1000000049
> base_000000050005BF8900000015 2021-09-22T17:39:19Z 000000050005BF8900000015
> ```
> 
> -- as we see here, it shows 3 backups for September 22, but in reality, we had only one (0000000500060895000000C3; two others are definitely older, see their LSNs)
> 
> I also see that the output of `backup-list` shows `modified` in the second column, not `created` – this seems to be inconsistent with the supposed order.
> 
> We need to ignore modtime and use either LSN or creation time, order by it, and present in the output explicitly.

Hi! Can you please also provide the output of the `wal-g backup-list --detail` command?

**Comment by NikolayS** [comment]
@usernamedt here it is (`sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list --detail | tail -10`):

```
name                          modified             wal_segment_backup_start start_time           finish_time          hostname    data_dir        pg_version start_lsn        finish_lsn       is_permanent
...
base_0000000500060413000000F9 2021-09-20T12:41:36Z 0000000500060413000000F9 2021-09-20T04:10:20Z 2021-09-20T12:41:35Z hostnameXXX /path/to/pgdata 120006     1693333688920552 1694422252556592 false
base_000000050006063800000080 2021-09-21T08:30:01Z 000000050006063800000080 2021-09-21T00:00:02Z 2021-09-21T08:30:00Z hostnameXXX /path/to/pgdata 120007     1695689596396096 1696320379778216 false
base_0000000500060895000000C3 2021-09-22T08:38:31Z 0000000500060895000000C3 2021-09-22T00:00:02Z 2021-09-22T08:38:30Z hostnameXXX /path/to/pgdata 120007     1698289175215536 1699026661799232 false
base_0000000500060ADB00000083 2021-09-23T08:38:11Z 0000000500060ADB00000083 2021-09-23T00:00:02Z 2021-09-23T08:38:10Z hostnameXXX /path/to/pgdata 120007     1700787784252472 1701710696775192 false
base_0000000500060D5B0000001D 2021-09-24T08:40:43Z 0000000500060D5B0000001D 2021-09-24T00:00:02Z 2021-09-24T08:40:42Z hostnameXXX /path/to/pgdata 120007     1703534855636720 1704097980979656 false
base_0000000500060F75000000C8 2021-09-25T08:40:48Z 0000000500060F75000000C8 2021-09-25T00:00:02Z 2021-09-25T08:40:47Z hostnameXXX /path/to/pgdata 120007     1705848401506504 1706147095008432 false
base_0000000500061039000000BC 2021-09-26T08:33:26Z 0000000500061039000000BC 2021-09-26T00:00:02Z 2021-09-26T08:33:25Z hostnameXXX /path/to/pgdata 120007     1706690014373488 1706986898397688 false
base_00000005000610EE000000E5 2021-09-27T08:40:38Z 00000005000610EE000000E5 2021-09-27T00:00:02Z 2021-09-27T08:40:37Z hostnameXXX /path/to/pgdata 120006     1707468093682336 1708190998879856 false
base_0000000500061353000000A9 2021-09-28T08:45:10Z 0000000500061353000000A9 2021-09-28T00:00:02Z 2021-09-28T08:45:09Z hostnameXXX /path/to/pgdata 120007     1710099898883816 1710808253201512 false
base_00000005000615B600000040 2021-09-29T08:44:02Z 00000005000615B600000040 2021-09-29T00:00:02Z 2021-09-29T08:44:01Z hostnameXXX /path/to/pgdata 120007     1712722372283792 1713393884340464 false
```

meanwhile, the output of `sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list | tail -10`:
```
name                          modified             wal_segment_backup_start
...
base_000000050005C4D9000000F0 2021-09-26T10:33:30Z 000000050005C4D9000000F0
base_00000005000610EE000000E5 2021-09-27T08:40:38Z 00000005000610EE000000E5
base_000000050005F68E0000009F 2021-09-27T13:10:13Z 000000050005F68E0000009F
base_000000050005C6E3000000B3 2021-09-27T13:42:11Z 000000050005C6E3000000B3
base_0000000500061353000000A9 2021-09-28T08:45:10Z 0000000500061353000000A9
base_000000050005C90B00000073 2021-09-28T09:49:05Z 000000050005C90B00000073
base_000000050005F8FB0000006E 2021-09-28T15:34:33Z 000000050005F8FB0000006E
base_00000005000615B600000040 2021-09-29T08:44:02Z 00000005000615B600000040
base_000000050005FB8000000089 2021-09-29T12:51:08Z 000000050005FB8000000089
base_000000050005CB30000000BD 2021-09-29T13:32:37Z 000000050005CB30000000BD
```

Something is off with the `modified` column.

**Comment by usernamedt** [comment]
> @usernamedt here it is (`sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list --detail | tail -10`):
> 
> 
> 
> ```
> 
> name                          modified             wal_segment_backup_start start_time           finish_time          hostname    data_dir        pg_version start_lsn        finish_lsn       is_permanent
> 
> ...
> 
> base_0000000500060413000000F9 2021-09-20T12:41:36Z 0000000500060413000000F9 2021-09-20T04:10:20Z 2021-09-20T12:41:35Z hostnameXXX /path/to/pgdata 120006     1693333688920552 1694422252556592 false
> 
> base_000000050006063800000080 2021-09-21T08:30:01Z 000000050006063800000080 2021-09-21T00:00:02Z 2021-09-21T08:30:00Z hostnameXXX /path/to/pgdata 120007     1695689596396096 1696320379778216 false
> 
> base_0000000500060895000000C3 2021-09-22T08:38:31Z 0000000500060895000000C3 2021-09-22T00:00:02Z 2021-09-22T08:38:30Z hostnameXXX /path/to/pgdata 120007     1698289175215536 1699026661799232 false
> 
> base_0000000500060ADB00000083 2021-09-23T08:38:11Z 0000000500060ADB00000083 2021-09-23T00:00:02Z 2021-09-23T08:38:10Z hostnameXXX /path/to/pgdata 120007     1700787784252472 1701710696775192 false
> 
> base_0000000500060D5B0000001D 2021-09-24T08:40:43Z 0000000500060D5B0000001D 2021-09-24T00:00:02Z 2021-09-24T08:40:42Z hostnameXXX /path/to/pgdata 120007     1703534855636720 1704097980979656 false
> 
> base_0000000500060F75000000C8 2021-09-25T08:40:48Z 0000000500060F75000000C8 2021-09-25T00:00:02Z 2021-09-25T08:40:47Z hostnameXXX /path/to/pgdata 120007     1705848401506504 1706147095008432 false
> 
> base_0000000500061039000000BC 2021-09-26T08:33:26Z 0000000500061039000000BC 2021-09-26T00:00:02Z 2021-09-26T08:33:25Z hostnameXXX /path/to/pgdata 120007     1706690014373488 1706986898397688 false
> 
> base_00000005000610EE000000E5 2021-09-27T08:40:38Z 00000005000610EE000000E5 2021-09-27T00:00:02Z 2021-09-27T08:40:37Z hostnameXXX /path/to/pgdata 120006     1707468093682336 1708190998879856 false
> 
> base_0000000500061353000000A9 2021-09-28T08:45:10Z 0000000500061353000000A9 2021-09-28T00:00:02Z 2021-09-28T08:45:09Z hostnameXXX /path/to/pgdata 120007     1710099898883816 1710808253201512 false
> 
> base_00000005000615B600000040 2021-09-29T08:44:02Z 00000005000615B600000040 2021-09-29T00:00:02Z 2021-09-29T08:44:01Z hostnameXXX /path/to/pgdata 120007     1712722372283792 1713393884340464 false
> 
> ```
> 
> 
> 
> meanwhile, the output of `sudo  /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list | tail -10`:
> 
> ```
> 
> name                          modified             wal_segment_backup_start
> 
> ...
> 
> base_000000050005C4D9000000F0 2021-09-26T10:33:30Z 000000050005C4D9000000F0
> 
> base_00000005000610EE000000E5 2021-09-27T08:40:38Z 00000005000610EE000000E5
> 
> base_000000050005F68E0000009F 2021-09-27T13:10:13Z 000000050005F68E0000009F
> 
> base_000000050005C6E3000000B3 2021-09-27T13:42:11Z 000000050005C6E3000000B3
> 
> base_0000000500061353000000A9 2021-09-28T08:45:10Z 0000000500061353000000A9
> 
> base_000000050005C90B00000073 2021-09-28T09:49:05Z 000000050005C90B00000073
> 
> base_000000050005F8FB0000006E 2021-09-28T15:34:33Z 000000050005F8FB0000006E
> 
> base_00000005000615B600000040 2021-09-29T08:44:02Z 00000005000615B600000040
> 
> base_000000050005FB8000000089 2021-09-29T12:51:08Z 000000050005FB8000000089
> 
> base_000000050005CB30000000BD 2021-09-29T13:32:37Z 000000050005CB30000000BD
> 
> ```
> 
> 
> 
> Sorting for the short version differs from that for the long one.

Ok, then it appears that WAL-G sorts the backups using the creation time only when being launched with the '--detail' flag. I think that this behavior difference is not cool and needs a fix. I can look into it in the nearby future.

**Comment by NikolayS** [comment]
@usernamedt right, this matches my understanding. Thanks.

**Comment by usernamedt** [comment]
> @usernamedt right, this matches my understanding. Thanks.

I've written a plan to implement the use of the backup creation time in the generic backup-list handler. Hopefully, I or somebody else will put an effort to implement it.

**Comment by rhenchen-gtlb** [comment]
I'd like to raise a higher concern regarding this old issue, as it can completely break the logic of incremental backups.

We noticed this recently in our environment when trying to implement incremental backups while also we have a retention policy to move backups older than 7 days into coldline storage, which updates the "modified" date of backup.

As we can notice bellow the 2 more recent incremental backups are a delta agains `000000020000B299000000FA` and  `000000020000B2A500000039` who were taken 7 days ago.

```
gitlab-psql@patroni-sec-v17-02-db-gstg.c.gitlab-staging-1.internal:~$ /usr/bin/envdir /etc/wal-g.d/env /opt/wal-g/bin/wal-g backup-list --detail --pretty
INFO: 2026/03/13 06:32:04.727780 List backups from storages: [default]
+----+----------------------------------------------------------+-----------------------------------+--------------------------+--------------+-----------------------------------+-----------------------------------+----------------------------+-----------------------------------+------------+---------------+---------------+-----------+
|  # | BACKUP NAME                                              | MODIFIED                          | WAL FILE NAME            | STORAGE NAME | START TIME                        | FINISH TIME                       | HOSTNAME                   | DATADIR                           | PG VERSION | START LSN     | FINISH LSN    | PERMANENT |
+----+----------------------------------------------------------+-----------------------------------+--------------------------+--------------+-----------------------------------+-----------------------------------+----------------------------+-----------------------------------+------------+---------------+---------------+-----------+
...
| 80 | base_000000020000B299000000FA                            | Wednesday, 11-Mar-26 07:58:09 UTC | 000000020000B299000000FA | default      | Tuesday, 03-Mar-26 00:00:03 UTC   | Tuesday, 03-Mar-26 00:06:03 UTC   | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B299/FA01A820 | B29A/9005C60  | false     |
| 81 | base_000000020000B29F00000099                            | Thursday, 12-Mar-26 18:38:00 UTC  | 000000020000B29F00000099 | default      | Wednesday, 04-Mar-26 00:00:03 UTC | Wednesday, 04-Mar-26 00:06:03 UTC | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B29F/99011480 | B29F/A40120F8 | false     |
| 82 | base_000000020000B2A500000039                            | Thursday, 12-Mar-26 18:38:00 UTC  | 000000020000B2A500000039 | default      | Thursday, 05-Mar-26 00:00:04 UTC  | Thursday, 05-Mar-26 00:06:03 UTC  | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2A5/39046C08 | B2A5/47000000 | false     |
| 83 | base_000000020000B2AA000000DA                            | Friday, 06-Mar-26 00:06:05 UTC    | 000000020000B2AA000000DA | default      | Friday, 06-Mar-26 00:00:04 UTC    | Friday, 06-Mar-26 00:06:04 UTC    | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2AA/DA000190 | B2AA/E50000D8 | false     |
| 84 | base_000000020000B2B00000007A                            | Saturday, 07-Mar-26 00:06:06 UTC  | 000000020000B2B00000007A | default      | Saturday, 07-Mar-26 00:00:04 UTC  | Saturday, 07-Mar-26 00:06:05 UTC  | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2B0/7A030788 | B2B0/860000C0 | false     |
| 85 | base_000000020000B2B600000019                            | Sunday, 08-Mar-26 00:06:05 UTC    | 000000020000B2B600000019 | default      | Sunday, 08-Mar-26 00:00:04 UTC    | Sunday, 08-Mar-26 00:06:04 UTC    | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2B6/19000128 | B2B6/230110D0 | false     |
| 86 | base_000000020000B2BB000000B8                            | Monday, 09-Mar-26 00:06:05 UTC    | 000000020000B2BB000000B8 | default      | Monday, 09-Mar-26 00:00:04 UTC    | Monday, 09-Mar-26 00:06:03 UTC    | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2BB/B8000028 | B2BB/C2000160 | false     |
| 87 | base_000000020000B2C100000056                            | Tuesday, 10-Mar-26 00:06:05 UTC   | 000000020000B2C100000056 | default      | Tuesday, 10-Mar-26 00:00:04 UTC   | Tuesday, 10-Mar-26 00:06:04 UTC   | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2C1/560204B8 | B2C1/61009CB8 | false     |
| 88 | base_000000020000B2C20000008C_D_000000020000B2C100000056 | Tuesday, 10-Mar-26 05:06:35 UTC   | 000000020000B2C20000008C | default      | Tuesday, 10-Mar-26 05:05:56 UTC   | Tuesday, 10-Mar-26 05:06:34 UTC   | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2C2/8C002028 | B2C2/92000090 | false     |
| 89 | base_000000020000B2C2000000D2_D_000000020000B2C20000008C | Tuesday, 10-Mar-26 06:21:02 UTC   | 000000020000B2C2000000D2 | default      | Tuesday, 10-Mar-26 06:20:27 UTC   | Tuesday, 10-Mar-26 06:21:01 UTC   | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2C2/D23238C0 | B2C2/DC016AC8 | false     |
| 90 | base_000000020000B2C2000000DC                            | Tuesday, 10-Mar-26 06:28:56 UTC   | 000000020000B2C2000000DC | default      | Tuesday, 10-Mar-26 06:22:56 UTC   | Tuesday, 10-Mar-26 06:28:55 UTC   | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2C2/DC0037B0 | B2C2/E40000C0 | false     |
| 91 | base_000000020000B2C6000000F7_D_000000020000B2C2000000DC | Wednesday, 11-Mar-26 00:00:42 UTC | 000000020000B2C6000000F7 | default      | Wednesday, 11-Mar-26 00:00:03 UTC | Wednesday, 11-Mar-26 00:00:40 UTC | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2C6/F70D2280 | B2C7/1012BB8  | false     |
| 92 | base_000000020000B2CC00000096_D_000000020000B299000000FA | Thursday, 12-Mar-26 00:00:49 UTC  | 000000020000B2CC00000096 | default      | Thursday, 12-Mar-26 00:00:03 UTC  | Thursday, 12-Mar-26 00:00:48 UTC  | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2CC/960F4D90 | B2CC/A0046820 | false     |
| 93 | base_000000020000B2D200000038_D_000000020000B2A500000039 | Friday, 13-Mar-26 00:00:48 UTC    | 000000020000B2D200000038 | default      | Friday, 13-Mar-26 00:00:04 UTC    | Friday, 13-Mar-26 00:00:46 UTC    | patroni-sec-v17-02-db-gstg | /var/opt/gitlab/postgresql/data17 | 170006     | B2D2/38001708 | B2D2/4100A830 | false     |
+----+----------------------------------------------------------+-----------------------------------+--------------------------+--------------+-----------------------------------+-----------------------------------+----------------------------+-----------------------------------+------------+---------------+---------------+-----------+
```

The full investigation can be found at https://gitlab.com/gitlab-com/gl-infra/data-access/dbo/dbo-issue-tracker/-/work_items/695#note_3156395414
